File: simpleApi/src/main/java/com/lifemmo/pl/simple/api/handler/SampleApiForRoom.java
package com.lifemmo.pl.simple.api.handler;

import com.lifemmo.pl.base.communicator.ServerInfo;
import com.lifemmo.pl.base.protocol.Packet;
import com.lifemmo.pl.base.protocol.ReplyPacket;
import com.lifemmo.pl.base.service.*;
import com.lifemmo.pl.base.service.api.annotation.Api;
import com.lifemmo.pl.base.service.api.annotation.ApiBackendHandler;
import com.lifemmo.pl.base.service.api.annotation.ApiHandler;
import com.lifemmo.pl.base.service.api.annotation.Init;
import com.lifemmo.pl.simple.Simple;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

@Slf4j
@Component
@Api
public class SampleApiForRoom {
    private SystemPanel systemPanel;
    private ApiBaseSender apiBaseSender;

    private String roomSvcId = "room";
    private String roomType = "simple";
    private int success = 0;
    private int fail = 1;

    @Init
    public void init(SystemPanel systemPanel,ApiBaseSender apiBaseSender){
        this.systemPanel = systemPanel;
        this.apiBaseSender = apiBaseSender;
    }

    @ApiHandler(msgName = "CreateRoomReq")
    public void createRoom(String sessionInfo, Packet packet, ApiSender apiSender) throws Exception{
        log.info("CreateRoom : sessionInfo:{},msgName:{}",sessionInfo,packet.getMsgName());

        String data = Simple.CreateRoomReq.parseFrom(packet.buffer()).getData();

        ServerInfo randRoomServerInfo  = systemPanel.randomServerInfo("room");
        String roomEndpoint = randRoomServerInfo.getBindEndpoint();

        var result = apiSender.createRoom(
                roomEndpoint,
                roomType,
                new Packet(Simple.CreateRoomAsk.newBuilder().setData(data).build()));

        var createRoomAnswer = Simple.CreateRoomAnswer.parseFrom(result.getCreateRoomRes().buffer());
        var roomId = result.getRoomId();

        if(result.isSuccess()){
            apiSender.reply(new ReplyPacket(
                    Simple.CreateRoomRes.newBuilder()
                            .setData(createRoomAnswer.getData())
                            .setRoomId(roomId)
                            .setRoomEndpoint(roomEndpoint)
                            .build())
            );
        }else{
            apiSender.reply(new ReplyPacket(result.getErrorCode()));
        }

    }

    @ApiHandler(msgName = "JoinRoomReq")
    public void joinRoom(String sessionInfo, Packet packet, ApiSender apiSender) throws Exception{

        log.info("joinRoom : sessionInfo:{},msgName:{}",sessionInfo,packet.getMsgName());
        var request = Simple.JoinRoomReq.parseFrom(packet.buffer());
        var data = request.getData();
        var roomId = request.getRoomId();
        var roomEndpoint = request.getRoomEndpoint();
        var accountId = Long.parseLong(sessionInfo);
        var sessionEndpoint = apiSender.sessionEndpoint();
        var sid = apiSender.sid();

        var result = apiSender.joinRoom(roomEndpoint,
                roomId,
                accountId,
                sessionEndpoint,
                sid,
                new Packet(Simple.JoinRoomAsk.newBuilder().setData(data).build())
        );

        if(result.isSuccess()){
            var joinRoomAnswer = Simple.JoinRoomAnswer.parseFrom(result.getJoinRoomRes().buffer());
            apiSender.reply(
                    new ReplyPacket(Simple.JoinRoomRes.newBuilder().setData(joinRoomAnswer.getData()).build())
            );
        }else{
            apiSender.reply(new ReplyPacket(result.getErrorCode()));
        }


    }

    @ApiHandler(msgName = "CreateJoinRoomReq")
    public void createJoinRoom(String sessionInfo, Packet packet, ApiSender apiSender) throws Exception{
        log.info("createJoinRoom : sessionInfo:{},msgName:{}",sessionInfo,packet.getMsgName());
        var request = Simple.CreateJoinRoomReq.parseFrom(packet.buffer());
        var data = request.getData();
        var roomId = request.getRoomId();
        var roomEndpoint = request.getRoomEndpoint();
        var accountId = Long.parseLong(sessionInfo);
        var sessionEndpoint = apiSender.sessionEndpoint();
        var sid = apiSender.sid();
        var createPayload = new Packet(Simple.CreateRoomAsk.newBuilder().setData(data).build());
        var joinPayload = new Packet(Simple.JoinRoomAsk.newBuilder().setData(data).build());

        var result = apiSender.createJoinRoom(
                roomEndpoint,roomType,roomId,createPayload,
                accountId,sessionEndpoint,sid, joinPayload);

        if(result.isSuccess()){
            var joinRoomAnswer = Simple.CreateJoinRoomAnswer.parseFrom(result.getJoinRoomRes().buffer());
            apiSender.reply(
                    new ReplyPacket(Simple.CreateJoinRoomRes.newBuilder().setData(joinRoomAnswer.getData()).build())
            );
        }else{
            apiSender.reply(new ReplyPacket(result.getErrorCode()));
        }
    }

    @ApiBackendHandler(msgName = "LeaveRoomNotify")
    public void leaveRoomNotify(String sessionInfo, Packet packet, ApiBackendSender apiBackendSender) throws Exception{
        log.info("leaveRoomNotify : sessionInfo:{},msgName:{}",sessionInfo,packet.getMsgName());
        BaseSender baseSender = ControlContext.INSTANCE.getBaseSender();
        var notify = Simple.LeaveRoomNotify.parseFrom(packet.buffer());
        baseSender.sendToClient(notify.getSessionEndpoint(),notify.getSid(),new Packet(notify));
    }

    @ApiBackendHandler(msgName="HelloToApiReq")
    public void helloToApiReq(String sessionInfo, Packet packet, ApiBackendSender apiBackendSender) throws Exception{
        log.info("helloToApiReq : sessionInfo:{},msgName:{}",sessionInfo,packet.getMsgName());
        String data = Simple.HelloToApiReq.parseFrom(packet.buffer()).getData();
        apiBackendSender.reply(new ReplyPacket(Simple.HelloToApiRes.newBuilder().setData(data).build()));

    }
}

File: simpleApi/src/main/kotlin/org/ulalax/playhouse/simple/api/handler/SampleApiForRoom.kt
package org.ulalax.playhouse.simple.api.handler

import com.google.protobuf.Descriptors.Descriptor
import org.apache.logging.log4j.kotlin.logger
import org.springframework.stereotype.Component
import org.ulalax.playhouse.communicator.message.Packet
import org.ulalax.playhouse.communicator.message.ReplyPacket
import org.ulalax.playhouse.service.*
import org.ulalax.playhouse.service.ControlContext.baseSender
import org.ulalax.playhouse.service.api.ApiService
import org.ulalax.playhouse.service.api.BackendHandlerRegister
import org.ulalax.playhouse.service.api.HandlerRegister
import org.ulalax.playhouse.simple.Simple.*
import org.ulalax.playhouse.simple.api.SpringContext

@Component
class SampleApiForRoom : ApiService {
    private val log = logger()

    private lateinit var systemPanel: SystemPanel
    private lateinit var sender: Sender
    private val roomSvcId = "room"
    private val roomType = "simple"
    private val success = 0
    private val fail = 1
    private val roomServiceId:Short = 3

    override suspend fun init(systemPanel: SystemPanel, sender: Sender) {
        this.systemPanel = systemPanel
        this.sender = sender
    }
    override fun handles(register: HandlerRegister,backendRegister: BackendHandlerRegister) {
        register.add(CreateRoomReq.getDescriptor().index,::createStage)
        register.add(JoinRoomReq.getDescriptor().index,::joinStage)
        register.add(CreateJoinRoomReq.getDescriptor().index,::createJoinStage)

        backendRegister.add(LeaveRoomNotify.getDescriptor().index,::leaveRoomNotify)
        backendRegister.add(HelloReq.getDescriptor().index,::helloToApiReq)
    }

    override fun instance(): ApiService {
        return SpringContext.getContext().getBean(this::class.java)
    }

    suspend fun createStage(packet: Packet, apiSender: ApiSender) {
        log.info("CreateRoom : accountId:${apiSender.accountId}," +
                "msgName:${ getDescriptor().messageTypes.find { it.index == packet.msgId }!!.name}")

        val data: String = CreateRoomReq.parseFrom(packet.data()).data
        val randRoomServerInfo = systemPanel.getServerInfoByService(roomServiceId)
        val roomEndpoint = randRoomServerInfo.bindEndpoint()
        val stageId = this.systemPanel.generateUUID()
        val result = apiSender.createStage(roomEndpoint, roomType,stageId, Packet(CreateRoomAsk.newBuilder().setData(data).build()))

        val createRoomAnswer = CreateRoomAnswer.parseFrom(result.createStageRes.data())

        log.info("stageId:$stageId")

        if (result.isSuccess()) {
            apiSender.reply(
                ReplyPacket(CreateRoomRes.newBuilder().setData(createRoomAnswer.data)
                        .setStageId(stageId)
                        .setPlayEndpoint(roomEndpoint)
                        .build()
                )
            )
        } else {
            apiSender.reply(ReplyPacket(result.errorCode))
        }
    }

    suspend fun joinStage( packet: Packet, apiSender: ApiSender) {
        log.info("joinRoom : accountId:${apiSender.accountId}, sid:${apiSender.sid}," +
                "msgName:${getDescriptor().messageTypes.find { it.index == packet.msgId }!!.name}")

        val request: JoinRoomReq = JoinRoomReq.parseFrom(packet.data())
        val data: String = request.data
        val stageId: Long = request.roomId
        val roomEndpoint: String = request.playEndpoint
        val result = apiSender.joinStage(
            roomEndpoint,
            stageId,
            Packet(JoinRoomAsk.newBuilder().setData(data).build())
        )
        if (result.isSuccess()) {
            log.info("stageIdx:${result.stageIndex}")

            val joinRoomAnswer = JoinRoomAnswer.parseFrom(result.joinStageRes.data())
            apiSender.reply(ReplyPacket(JoinRoomRes.newBuilder().setData(joinRoomAnswer.data).setStageIdx(result.stageIndex).build()))
        } else {
            apiSender.reply(ReplyPacket(result.errorCode))
        }
    }

    suspend fun createJoinStage(packet: Packet, apiSender: ApiSender) {
        log.info("CreateJoinRoomReq : accountId:${apiSender.accountId},sid:${apiSender.sid}" +
                "msgName:${getDescriptor().messageTypes.find { it.index == packet.msgId }!!.name}")

        val request: CreateJoinRoomReq = CreateJoinRoomReq.parseFrom(packet.data())
        val data = request.data
        val stageId: Long = request.roomId
        val roomEndpoint: String = request.playEndpoint
        val createPayload = Packet(CreateRoomAsk.newBuilder().setData(data).build())
        val joinPayload = Packet(JoinRoomAsk.newBuilder().setData(data).build())
        val result = apiSender.createJoinStage(
            roomEndpoint,
            roomType,
            stageId,
            createPayload,
            joinPayload
        )
        if (result.isSuccess()) {
            val joinRoomAnswer = CreateJoinRoomAnswer.parseFrom(result.joinStageRes.data())
            apiSender.reply(
                ReplyPacket(CreateJoinRoomRes.newBuilder().setData(joinRoomAnswer.data).setStageIdx(result.stageIndex).build())
            )
        } else {
            apiSender.reply(ReplyPacket(result.errorCode))
        }
    }

    suspend fun leaveRoomNotify(packet: Packet, apiBackendSender: ApiBackendSender) {
        log.info("leaveRoomNotify : accountId:${apiBackendSender.accountId}" +
                "msgName:${getDescriptor().messageTypes.find { it.index == packet.msgId }!!.name}")

        val baseSender = baseSender
        val notify = LeaveRoomNotify.parseFrom(packet.data())
        baseSender.sendToClient(notify.sessionEndpoint, notify.sid, Packet(notify))
    }

    suspend fun helloToApiReq(packet: Packet, apiBackendSender: ApiBackendSender) {
        log.info("helloToApiReq : accountId:${apiBackendSender.accountId}" +
                "msgName:${getDescriptor().messageTypes.find { it.index == packet.msgId }!!.name}")

        val data: String = HelloToApiReq.parseFrom(packet.data()).data
        apiBackendSender.reply(ReplyPacket(HelloToApiRes.newBuilder().setData(data).build()))
    }


}